src/client/AM.PA.MacOs/PersonalAnalytics/App/AppDelegate.swift

```diff
@@ -280,10 +301,47 @@ class AppDelegate: NSObject, NSApplicationDelegate, NSUserNotificationCenterDele
 
     
     func userNotificationCenter(_ center: NSUserNotificationCenter, didActivate notification: NSUserNotification) {
-        //print("Using delelgate for NSUsernotification")
-        self.toggleSummary(notification.self)
+
+        if notification.title == "What are you feeling?" {
+            // EMOTION POP-UP ACTIVATION
+            // If the notification is postponed...
+            if let choosen = notification.additionalActivationAction, let actionIdentifier = choosen.identifier {
+                switch actionIdentifier {
+                case "5m":
+                    emotionPopUpController.emotionTracker.scheduleNotification(minutesSinceNow: 5*60)
+                    print("Notification postponed. It will display 5 minutes from now!")
+                case "30m":
+                    emotionPopUpController.emotionTracker.scheduleNotification(minutesSinceNow: 30*60)
+                    print("Notification postponed. It will display 30 minutes from now!")
+                case "1h":
+                    emotionPopUpController.emotionTracker.scheduleNotification(minutesSinceNow: 60*60)
+                    print("Notification postponed. It will display 60 minutes from now!")
+                default:
+                    print("Something went wrong: UserNotification additionalActivationAction not recognized")
+                }
+            } else {
+                // If the notification is accepted...
+
+                // Shows EmotionPopUp
+                self.showEmotionPopUp(self)
+                // Removes the notification from the NotificationCenter
+                center.removeDeliveredNotification(notification)
+            }
+        } else {
+            // SUMMARY ACTIVATION
+            //print("Using delelgate for NSUsernotification")
+            self.toggleSummary(notification.self)
+        }
+
+
+
+
     }
-    
+
+    func userNotificationCenter(_ center: NSUserNotificationCenter, shouldPresent notification: NSUserNotification) -> Bool {
```

Comment:
`userNotificationCenter(_:shouldPresent:)` is an instance method of the class `NSUserNotificationCenterDelegate` that is called when the user notification center has decided not to present a scheduled notification.

The function returns true if the user notification should be displayed regardless of the notification center’s decision.  I’m quite new to Mac OS programming, so I may have misunderstood this. In my idea, what I do is force the notification to always be shown, since its regular operation it’s critical to my experiment. I noticed indeed that when I omit these lines, some notifications “get lost”: they are never presented to the user.

I will examine this issue in more depth as soon as possible. Meanwhile, I delete the function from the AppDelegate.